Sources/DeepLearning/Layer.swift

```diff
@@ -500,3 +500,28 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
         return upSampling.reshaped(toShape: upSampledShape)
     }
 }
+
+@_fixed_layout
+public struct Flatten<Scalar: TensorFlowFloatingPoint>: Layer {
+    @differentiable(wrt: (self, input))
+    public func applied(to input: Tensor<Scalar>, in _: Context) -> Tensor<Scalar> {
+        let batchSize = input.shape[0]
```

Comment:
PR [#22848](https://github.com/apple/swift/pull/22848) changed the behavior of differentiation through non-differentiable types, so it may break this function (requiring a `.withoutDerivative()` call somewhere). Holding off until the next toolchain release (Monday) will be a good idea.

Also, a more efficient implementation would be to use `.shapeTensor` whenever you need to get the shape from a tensor, and not use any values of `TensorShape` type in the function body. One line 510, `newShape` can be defined using `Tensor`'s concatenation operator, `Tensor.concatenated(with:)` or `++`. (BTW, concatenation ops do not have a derivative defined yet, would you like to add one?)

Reply:
New toolchain released!

Reply:
My recommendation would be:
```swift
public func applied(to input: Tensor<Scalar>, in _: Context) -> Tensor<Scalar> {
    let batchSize = input.shape[0]
    let remaining = input.shape[1..<input.rank].contiguousSize
    return input.reshaped(to: [batchSize, remaining])
}
```
But I see also @rxwei's point about preferring purely TF-side shape manipulation.

If we don't switch to that completely, though, we should replace `reshaped(toShape: Tensor<Int32>(foo))` with `reshaped(to: foo)` (assuming that actually works now).

Reply:
The following is the host-free version.

```swift
public func applied(to input: Tensor<Scalar>, in _: Context) -> Tensor<Scalar> {
    let batchSize = input.shapeTensor[0]
    let remaining = input.shapeTensor[1..<input.rank].product()
    return input.reshaped(toTensor: Tensor([batchSize, remaining]))
}
```

Reply:
The host-free version is currently an order of magnitude slower on CPU than the hosty version (about 30us for mine and 300us for yours).

Reply:
D'oh that's terrible. Ok.